File: leetCode_Python/prob22.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""

"""

# Name: 22. Generate Parentheses
# Website: https://leetcode.com/problems/generate-parentheses/
# Description:
import sys


class Solution:
    def generateParenthesis(self, N):
        ans = []
        def backtrack(s="", l=0, r=0):
            if len(s) == 2 * N:
                ans.append(s)
            if l < N:
                backtrack(s+"(", l+1, r)
            if r < l:
                backtrack(s+")", l, r+1)
        backtrack()
        return ans

if __name__ == "__main__":
    pass

File: leetcode/prob22.cpp
#include <iostream>
#include <string>
#include <vector>

#include "gtest/gtest.h"

using namespace std;

// Name: 22. Generate Parentheses
// Website: https://leetcode.com/problems/generate-parentheses/
// Description:

class Solution {
public:
    std::vector<std::string> generateParenthesis(int n) {
        std::vector<std::string> res;
        addingpar(res, "", n, 0);
        return res;
    }
    void addingpar(std::vector<std::string> &v, std::string str, int n, int m){
        if (n==0 && m==0) {
            v.push_back(str);
            return;
        }

        if (n > 0) {
            addingpar(v, str+"(", n-1, m+1);
        }

        if (m > 0) {
            addingpar(v, str+")", n, m-1);
        }
    }
};

Solution solu;

// TEST(Fuction, one) {}